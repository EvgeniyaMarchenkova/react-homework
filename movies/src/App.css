.App {
  text-align: center;
}

.selected {
  font-weight: bold;
}
