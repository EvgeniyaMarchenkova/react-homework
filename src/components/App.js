import React, {Component} from "react";

class App extends React.Component {
    render() {
        return <div>My component</div>
    }
}

export default App;