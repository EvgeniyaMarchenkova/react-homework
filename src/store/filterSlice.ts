import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from './store';
import { SearchBy } from '../model';

interface FilterState {
  search: string;
  searchBy: SearchBy;
}

const initialState: FilterState = {
  search: '',
  searchBy: null,
};

export const filterSlice = createSlice({
  name: 'filter',
  initialState,
  reducers: {
    setSearch: (state: FilterState, action: PayloadAction<string>) => {
      state.search = action.payload;
    },
    setSearchBy: (state: FilterState, action: PayloadAction<SearchBy>) => {
      state.searchBy = action.payload;
    },
  },
});

export const { setSearch, setSearchBy } = filterSlice.actions;

export const selectSearch = (state: RootState) => state.filter.search;
export const selectSearchBy = (state: RootState) => state.filter.searchBy;

export default filterSlice.reducer;
