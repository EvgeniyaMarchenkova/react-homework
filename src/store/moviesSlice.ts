import { createSlice } from '@reduxjs/toolkit';
import type { PayloadAction } from '@reduxjs/toolkit';
import type { RootState } from './store';
import { Genre, MovieData } from '../model';

interface MoviesState {
  list: MovieData[];
  selectedMovie: MovieData;
}

const initialState: MoviesState = {
  list: [],
  selectedMovie: null,
};

export const moviesSlice = createSlice({
  name: 'movies',
  initialState,
  reducers: {
    setMovies: (state: MoviesState, action: PayloadAction<MovieData[]>) => {
      state.list = action.payload;
    },
    setSelectedMovie: (
      state: MoviesState,
      action: PayloadAction<MovieData>,
    ) => {
      state.selectedMovie = action.payload;
    },
  },
});

export const { setSelectedMovie, setMovies } = moviesSlice.actions;

// Other code such as selectors can use the imported `RootState` type
export const selectMovies = (state: RootState) => state.movies.list;
export const selectSelectedMovie = (state: RootState) =>
  state.movies.selectedMovie;

export default moviesSlice.reducer;
