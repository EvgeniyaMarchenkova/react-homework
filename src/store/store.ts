import { configureStore } from '@reduxjs/toolkit';
import moviesReducer from './moviesSlice';
import filterReducer from './filterSlice';
import { moviesApi } from '../query/movies';

export const store = configureStore({
  reducer: {
    movies: moviesReducer,
    filter: filterReducer,
    [moviesApi.reducerPath]: moviesApi.reducer,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(moviesApi.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
