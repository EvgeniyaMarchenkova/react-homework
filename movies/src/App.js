
import './App.css';
import Counter from './components/Counter';
import Search from './components/Search';

const genres = ['ALL', 'DOCUMENTARY', 'COMEDY', 'HORROR', 'CRIME'];

function GenresSwitcher() {
  return (
    <div className="App">
      <header className="App-header">
        APP
      </header>
    </div>
  );
}

function App() {
  return (
    <div className="App">
      <header className="App-header">
        <GenresSwitcher />
        <Search/>
        <Counter/>
      </header>
    </div>
  );
}

export default App;