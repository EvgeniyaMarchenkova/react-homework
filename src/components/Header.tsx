import React, { ChangeEvent } from 'react';
import styled from 'styled-components';
import { resetSelectedMovie } from '../store/moviesSlice';
import { useAppDispatch } from '../store/hooks';

interface HeaderData {
  searchText: string;
  // onChangedSearchText: React.Dispatch<React.SetStateAction<string>>;
  openAddMovieWindow: React.Dispatch<React.SetStateAction<string>>;
}

const Input = styled.input`
  padding: 0.5em;
  margin: 0.5em;
  border: none;
  border-radius: 3px;
`;

const AddMovieButton = styled.button`
  position: absolute;
  top: 20px;
  right: 20px;
`;

const HeaderContainer = styled.div`
  height: 150px;
  line-height: 150px;
  background: -webkit-linear-gradient(
      45deg,
      #292929 25%,
      transparent 25%,
      transparent 75%,
      #292929 75%
    ),
    -webkit-linear-gradient(
        45deg,
        #292929 25%,
        transparent 25%,
        transparent 75%,
        #292929 75%
      ) 0.1875em 0.1875em,
    -webkit-radial-gradient(at 50% 0, #484847, #090909);
  background: linear-gradient(
      45deg,
      #292929 25%,
      transparent 25%,
      transparent 75%,
      #292929 75%
    ),
    linear-gradient(
        45deg,
        #292929 25%,
        transparent 25%,
        transparent 75%,
        #292929 75%
      )
      0.1875em 0.1875em,
    radial-gradient(at 50% 0, #484847, #090909);
  background-size: 0.375em 0.375em, 0.375em 0.375em, 100% 100%;
`;

const Header = (props: HeaderData) => {
  const dispatch = useAppDispatch();

  return (
    <HeaderContainer>
      <AddMovieButton
        onClick={() => {
          dispatch(resetSelectedMovie());
          props.openAddMovieWindow(undefined);
        }}
      >
        + ADD MOVIE
      </AddMovieButton>
      <Input
        value={props.searchText}
        // onChange={
        // (ev: ChangeEvent<HTMLInputElement>) =>
        // props.onChangedSearchText(ev.target.value)
        // }
        type="text"
        placeholder="Select"
      />
      <button
      // onClick={() => props.onChangedSearchText('')}
      >
        Reset Search
      </button>
    </HeaderContainer>
  );
};

export default Header;
