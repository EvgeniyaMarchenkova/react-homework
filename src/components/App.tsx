import React, { useState, useEffect } from 'react';
import Header from './Header';
import MainContent from './MainContent';
import Footer from './Footer';
import ErrorBoundary from './ErrorBoundary';
import { ModalWindowType } from './../model';
import ModalWindowOpener from './ModalWindowOpener';
import { Genre } from '../model';

const INITIAL_MOVIE_LIST = [
  {
    value: 'DOCUMENTARY 1',
    genre: Genre.Documentary,
  },
  {
    value: 'COMEDY 1',
    genre: Genre.Comedy,
  },
  {
    value: 'DOCUMENTARY 2',
    genre: Genre.Documentary,
  },
  {
    value: 'DOCUMENTARY 3',
    genre: Genre.Documentary,
  },
  {
    value: 'COMEDY 4',
    genre: Genre.Comedy,
  },
  {
    value: 'CRIME 2',
    genre: Genre.Crime,
  },
];

export interface Handlers {
  addMovie: Function;
  deleteMovie: Function;
  updateMovie: Function;
}

const App = () => {
  const [searchText, setSearchText] = useState('');
  const [modalWindowType, setModalWindowType] = useState(ModalWindowType.None);
  const [movies, setMovies] = useState([]);
  const [selectedMovie, setSelectedMovie] = useState('');

  useEffect(() => {
    setMovies(INITIAL_MOVIE_LIST);
  }, []);

  let MOVIE_HANDLERS = {
    delete: () => {
      console.log('delete', selectedMovie);
      setMovies(movies.filter((movie) => movie.value !== selectedMovie));    
    },
    add: () => {
      console.log('add');
    },
    edit: () => {
      console.log('update', selectedMovie);
    },
  };

  return (
    <>
      {modalWindowType !== ModalWindowType.None && (
        <ModalWindowOpener
          movieHandlers={MOVIE_HANDLERS}
          type={modalWindowType}
          onCloseWindow={() => setModalWindowType(ModalWindowType.None)}
        />
      )}
      <Header
        searchText={searchText}
        onChangedSearchText={setSearchText}
        openAddMovieWindow={() => setModalWindowType(ModalWindowType.AddMovie)}
      />
      <ErrorBoundary>
        <MainContent
          movies={movies}
          searchText={searchText}
          openModalWindow={(type: ModalWindowType) => setModalWindowType(type)}
          selectMovie={setSelectedMovie}
        />
      </ErrorBoundary>
      <Footer />
    </>
  );
};

export default App;
