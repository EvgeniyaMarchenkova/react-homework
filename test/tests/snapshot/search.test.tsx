import React from "react";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import Header from "../../../src/components/Header";
import App from "../../../src/components/App";
import { Provider } from "react-redux";
import { store } from "../../../src/store/store";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";

describe("Search should", () => {
  it("render correctly", () => {
    const fakeHandler = jest.fn();
    const header = renderer
      .create(
        <Provider store={store}>
          <Header
            onChangedSearchText={fakeHandler}
            openAddMovieWindow={fakeHandler}
          >
            Facebook
          </Header>
        </Provider>
      )
      .toJSON();
    expect(header).toMatchSnapshot();
  });

  it("contain input for search", async () => {
    const fakeHandlerChangeInput = jest.fn();
    const fakeHandlerAddMovieClick = jest.fn();
    render(
      <Provider store={store}>
        <Header
          onChangedSearchText={fakeHandlerChangeInput}
          openAddMovieWindow={fakeHandlerAddMovieClick}
        >
          Facebook
        </Header>
      </Provider>
    );
    const input = await screen.getByRole("textbox");
    const input2 = await screen.getByPlaceholderText("Select");
    expect(input).toBeInTheDocument();
    // await fireEvent.change(input2, { target: { value: "dragon" } });
    // expect((input2 as HTMLInputElement).value).toEqual("dragon");

    // await userEvent.type(input, "dragon");
    // screen.debug();
    // expect(screen.getByRole("textbox") as HTMLInputElement).toHaveValue(
    //   "dragon"
    // );
  });
});

it("should invoke corresponding handler when user input value changes", async () => {
  const fakeHandlerChangeInput = jest.fn();
  const fakeHandlerAddMovieClick = jest.fn();
  render(
    <Provider store={store}>
      (
      <Header
        onChangedSearchText={fakeHandlerChangeInput}
        openAddMovieWindow={fakeHandlerAddMovieClick}
      >
        ) as any Facebook
      </Header>
    </Provider>
  );
  const input = await screen.getByRole("textbox");
  await fireEvent.change(input, { target: { value: "dragon" } });
  expect(fakeHandlerChangeInput).toHaveBeenCalledTimes(1);

  // await userEvent.type(input, "dragon");
  // screen.debug();
  // expect(screen.getByRole("textbox") as HTMLInputElement).toHaveValue(
  //   "dragon"
  // );
});
