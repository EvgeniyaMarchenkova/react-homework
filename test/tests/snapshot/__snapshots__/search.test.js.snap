// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly 1`] = `
<div
  className="sc-dkrFOg vuVxh"
>
  <button
    className="sc-gswNZR bkHYVn"
    onClick={[Function]}
  >
    + ADD MOVIE
  </button>
  <input
    className="sc-bcXHqe gAHLJq"
    onChange={[Function]}
    placeholder="Select"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Reset Search
  </button>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="sc-dkrFOg vuVxh"
>
  <button
    className="sc-gswNZR bkHYVn"
    onClick={[Function]}
  >
    + ADD MOVIE
  </button>
  <input
    className="sc-bcXHqe gAHLJq"
    onChange={[Function]}
    placeholder="Select"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Reset Search
  </button>
</div>
`;

exports[`renders correctly 2`] = `
<div
  className="sc-dkrFOg vuVxh"
>
  <button
    className="sc-gswNZR bkHYVn"
    onClick={[Function]}
  >
    + ADD MOVIE
  </button>
  <input
    className="sc-bcXHqe gAHLJq"
    onChange={[Function]}
    placeholder="Select"
    type="text"
    value=""
  />
  <button
    onClick={[Function]}
  >
    Reset Search
  </button>
</div>
`;
